cmd/connector_ethereum.go

```diff
@@ -50,35 +50,53 @@ func connectChain(
 	// for the TBTCSystem address from now on; default to Extensions.TBTC and
 	// warn if the ContractAddresses version is not set yet.
 	_, exists := config.Ethereum.ContractAddresses[ethereum.TBTCSystemContractName]
-	if !exists && len(config.Extensions.TBTC.TBTCSystem) != 0 {
+	if len(config.Extensions.TBTC.TBTCSystem) != 0 {
 		logger.Warn(
 			"TBTCSystem address configuration in Extensions.TBTC.TBTCSystem " +
 				"is DEPRECATED and will be removed. Please configure the " +
 				"TBTCSystem address alongside BondedECDSAKeep under " +
 				"Ethereum.ContractAddresses.",
 		)
-		config.Ethereum.ContractAddresses[ethereum.TBTCSystemContractName] =
-			config.Extensions.TBTC.TBTCSystem
 
-		// Flag that the contract address entry now exists to skip the next
-		// default.
-		exists = true
+		if !exists {
+			config.Ethereum.ContractAddresses[ethereum.TBTCSystemContractName] =
+				config.Extensions.TBTC.TBTCSystem
+
+			// Flag that the contract address entry now exists to skip the next
+			// default.
+			exists = true
+		} else {
+			if config.Ethereum.ContractAddresses[ethereum.TBTCSystemContractName] !=
+				config.Extensions.TBTC.TBTCSystem {
+				panic(
+					"Configured TBTCSystem contract and Extensions.TBTC.TBTCSystem " +
+						"do not match. Failing to boot to avoid misconfiguration. " +
+						"Please ensure ethereum.ContractAddresses." +
+						ethereum.TBTCSystemContractName + "is set to the correct " +
+						"tBTC system contract and remove Extensions.TBTC.TBTCSystem " +
+						"entry, then try starting again.",
+				)
+			}
+		}
 	}
 
 	// DEPRECATED: config.Ethereum.ContractAddresses is the correct container
 	// for the TBTCSystem address from now on; read SanctionedApplications and
 	// assume it has a single entry that is TBTCSystem, warn if
 	// SanctionedApplications needs to be used.
 	applicationAddresses := config.SanctionedApplications.AddressesStrings
-	if !exists && len(applicationAddresses) == 1 {
+	if len(applicationAddresses) == 1 {
```

Comment:
I'm not sure why did we have the difference between this check (`len(applicationAddresses) == 0`) and the next one (L102: `len(applicationAddresses) > 0`). Shoudln't we align them both to be `len(applicationAddresses) > 0`?

Reply:
I think my thinking was that only with a 1-length application address can we have high confidence that the address in question is the TBTCSystem address, and thus warn. Without it, we don't know what they've thrown in here.

That said, I also think I overthought it haha. Aligning makes sense. The only additional sanctioned application we might expect is the original tBTC deploy, but treating that as equally problematic from a config perspective makes sense.